core/src/main/java/com/graphhopper/routing/util/parsers/OSMCyclewayParser.java

```diff
@@ -13,81 +13,122 @@
+
+    public OSMCyclewayParser() {
+        this(new EnumEncodedValue<>(Cycleway.KEY, Cycleway.class, true));
     }
-    Cycleway cyclewayForward = Cycleway.find(readerWay.getTag("cycleway:" + drivingSide.toString()));
-    Cycleway cyclewayBackward = Cycleway.find(readerWay.getTag("cycleway:" + DrivingSide.reverse(drivingSide).toString()));
 
-    if (cycleway != Cycleway.MISSING) {
-      cyclewayEnc.setEnum(false, edgeFlags, cycleway);
-      cyclewayEnc.setEnum(true, edgeFlags, cycleway);
+    public OSMCyclewayParser(EnumEncodedValue<Cycleway> cyclewayEnc) {
+        this.cyclewayEnc = cyclewayEnc;
+        oneways.add("yes");
+        oneways.add("true");
+        oneways.add("1");
+        oneways.add("-1");
+
+        oppositeLanes.put("opposite", Cycleway.SHARED_LANE);
+        oppositeLanes.put("opposite_lane", Cycleway.LANE);
+        oppositeLanes.put("opposite_track", Cycleway.TRACK);
+        oppositeLanes.put("opposite_share_busway", Cycleway.SHARE_BUSWAY);
     }
-    if (cyclewayForward != Cycleway.MISSING) {
-      cyclewayEnc.setEnum(false, edgeFlags, cyclewayForward);
-      if (isOneway(readerWay)) {
-        cyclewayEnc.setEnum(true, edgeFlags, cyclewayForward);
-      }
+
+    @Override
+    public void createEncodedValues(EncodedValueLookup lookup, List<EncodedValue> list) {
+        list.add(cyclewayEnc);
     }
-    if (cyclewayBackward != Cycleway.MISSING) {
-      cyclewayEnc.setEnum(true, edgeFlags, cyclewayBackward);
-      if (isOneway(readerWay)) {
-        cyclewayEnc.setEnum(false, edgeFlags, cyclewayBackward);
-      }
+
+    @Override
+    public IntsRef handleWayTags(IntsRef edgeFlags, ReaderWay readerWay, IntsRef relationFlags) {
+        // Figuring out what kind of bike infra applies to the forward and backward direction
+        // is complicated.
+        //
+        // The goal here is, given the cycleway, cycleway:both:*, and cycleway:left/right:* tags
+        // and the driving side, to come up with a notion of forward and backward cycleway values,
+        // normalized not to use the "opposite" values.
+        //
+        // To simplify the number of cases, cycleway:both is treated as a shorthand for setting
+        // both cycleway:left and cycleway:right, instead of being treated as a different case.
+        // Note that this is different from how the cyclewag tag is treated (see
+        // https://wiki.openstreetmap.org/wiki/Key:cycleway:both).
+
+        Cycleway cycleway = Cycleway.find(readerWay.getTag("cycleway"));
+
+        // Default forward and backward to the "cycleway" value until we find a more
+        // specific value.
+        Cycleway cyclewayForward = cycleway, cyclewayBackward = cycleway;
+
+        // If we find opposite_* tags, normalize them, and treat that as the final word
+        // (discounting any contradictory left/right tags which would be a tagging error).
+        if (oppositeLanes.containsKey(cycleway)) {
+            cyclewayForward = Cycleway.SHARED_LANE;
```

Comment:
why do we set `cyclewayForward` to `Cycleway.SHARED_LANE`. shouldn't it keep its old value of `Cycleway.find(readerWay.getTag("cycleway"))`

Reply:
Because it doesn't make sense to say, "In the forward direction, the bike infra consists of a bike lane in the opposite direction." So I normalize it like the comment says.

See the documentation for the "opposite" values, starting with why they're discouraged: https://wiki.openstreetmap.org/wiki/Key:cycleway#Problems_with_opposite*_values

Reply:
that's fine, just wondering why `Cycleway.SHARED_LANE` is the default. could lead to incorrect behavior? unless this value is ignored in which case maybe `Cycleway.MISSING` is a better value? maybe im misunderstanding

Reply:
That's my interpretation of what the value is intended to be in the forward direction when the opposite values are used. Do you have a different interpretation of the tag docs, and on what basis?

Reply:
ah i see, i was misunderstanding `cycleway=opposite*`. objection retracted. for my information, how does the local `oppositeLanes` variable work? what info does it hold?

Reply:
should this be `oppositeLanes.get(cycleway)` instead so that `opposite_share_busway` can become `Cycleway.SHARE_BUSWAY` etc? or is the intention to flatten everything to `Cycleway.SHARED_LANE`

Reply:
`opposite_share_busway` has [no documentation and 99 uses](https://wiki.openstreetmap.org/wiki/Key:cycleway=opposite_share_busway) in the entire world and has been deprecated for years. We are way in the weeds even having this conversation. But my interpretation is that no, "You can bike on this otherwise one-way road by sharing the contraflow bus lane" does not imply "Also, there's a bus lane in the forward direction that you should also use."

Reply:
im pretty unclear on what `cycleway=opposite_track` etc means for the forward direction. i guess the options are make forwards and backwards the same, or pick a default value for forwards. my vote for a default value is still `Cycleway.MISSING`, because we don't have any information on what the forward direction